Improve parallel raster episode (#136)
carpentries-incubator/geospatial-python: _episodes/11-parallel-raster-computations.md

Being able to profile the computational time is thus essential, and we will see how to do that in a Jupyter environment
in the next section.

The example that we consider here is the application of a median filter to a satellite image.

> [Median filtering](https://en.wikipedia.org/wiki/Median_filter) is a common noise removal technique, and it consists in

Reviewer: ```suggestion
[Median filtering](https://en.wikipedia.org/wiki/Median_filter) is a common noise removal technique which
```

> replacing a pixel intensity with the value computed from the median of the surrounding pixels.

Reviewer: ```suggestion
replaces a pixel's value with the median value computed from its surrounding pixels.
```

~~~
{: .language-python}

We select the last scene, and extract the URL of the true-color image ("visual"):

~~~
assets = items[-1].assets  # last item's assets
visual_href = assets["visual"].href  # true color image
~~~
{: .language-python}

The true-color image is available as a raster file with 10 m resolution and 3 bands (you can verify this by opening the
file with `rioxarray`), which makes it a relatively large file (few hundreds MBs). In order to keep calculations
"manageable" (reasonable execution time and memory usage) we select here a lower resolution version of the image, taking
advantage of the so-called "pyramidal" structure of cloud-optimized GeoTIFFs (COGs). COGs, in fact, typically include

> multiple lower-resolution versions of the original image, called "overviews", in the same file. This allows to avoid

Reviewer: ```suggestion
multiple lower-resolution versions of the original image, called "overviews", in the same file. This allows us to avoid
```

of 4, and so on. Here, we open the third level overview (zoom factor 8) and check that the resolution is about 80 m:

~~~
import rioxarray
visual = rioxarray.open_rasterio(visual_href, overview_level=2)
print(visual.rio.resolution())
~~~
{: .language-python}

~~~
(79.97086671522214, -79.97086671522214)
~~~
{: .output}

Let's make sure the data has been loaded into memory before proceeding to time profile our raster calculation. Calling
the `.load()` method of a DataArray object triggers data loading:

~~~
visual = visual.load()
~~~
{: .language-python}

Note that by default data is loaded using Numpy arrays as underlying data structure. We can visualize the raster:

~~~
visual.plot.imshow(figsize=(10,10))
~~~
{: .language-python}

<img src="../fig/E11-01-true-color-image.png" title="Scene true color image" alt="true color image scene" width="612" style="display: block; margin: auto;" />

Let's now apply a median filter to the image while keeping track of the execution time of this task. The filter is
carried out in two steps: first, we define the size and centering of the region around each pixel that will be

> considered for the median calculation (the so-called "windows"), using the `.rolling()` method. We chose here windows

Reviewer: ```suggestion
considered for the median calculation (the so-called "windows"), using the `.rolling()` method. We choose here windows
```
